### DISCUSSION TITLE
Custom implementation of claim check

### DISCUSSION THREAD

**Discussion by rcdailey** [rebus-org/Rebus#1182] [Q&A]
I see libraries like Rebus and MassTransit implement their own version of the Claim Check pattern. However, all of these typically involve modifying the DTO structure to support claim check, which in my view defeats the purpose. I believe claim check should be transient -- if a *serialized* payload size is greater than some configurable threshold, then instead of being sent as the message body, it should be stored in some configurable data store (e.g. S3, DynamoDB) and a claim check ID header should be added to the message. 

On the receiving end, if the claim check header exists, pull data from that storage layer and deserialize it into the final strongly-typed DTO structure the consumer wants.

Because I do not see any available implementations that do claim check transparently like this, I'd like to implement my own. Could you provide some pointers/direction on the best way to extend Rebus with this functionality?

**Comment by mookid8000** [comment]
Sorry for being so slow to reply here.

Rebus basically supports two types of claim check out of the box via its "data bus" feature:

1. Manual mode where you create attachments yourself
2. Transparent mode where serialized message payloads above a certain threshold are transferred using the data bus 

You can read more about both here: https://github.com/rebus-org/Rebus/wiki/Data-bus

If you still prefer to build your own, it's totally possible (and probably also a fun exercise) to do so. I suggest you take a look at Rebus' existing implementation for hints on how it can be integrated.
